source/extensions/filters/http/ip_tagging/ip_tagging_filter.cc

```diff
@@ -13,19 +14,184 @@ namespace Extensions {
 namespace HttpFilters {
 namespace IpTagging {
 
+IpTagsProvider::IpTagsProvider(const envoy::config::core::v3::DataSource& ip_tags_datasource,
+                               IpTagsLoader& tags_loader, uint64_t ip_tags_refresh_interval_ms,
+                               IpTagsReloadSuccessCb reload_success_cb,
+                               IpTagsReloadErrorCb reload_error_cb, Event::Dispatcher& dispatcher,
+                               Api::Api& api, ThreadLocal::SlotAllocator& tls,
+                               Singleton::InstanceSharedPtr owner, absl::Status& creation_status)
+    : ip_tags_path_(ip_tags_datasource.filename()), tags_loader_(tags_loader),
+      time_source_(api.timeSource()),
+      ip_tags_refresh_interval_ms_(std::chrono::milliseconds(ip_tags_refresh_interval_ms)),
+      reload_success_cb_(reload_success_cb), reload_error_cb_(reload_error_cb), owner_(owner) {
+  RETURN_ONLY_IF_NOT_OK_REF(creation_status);
+  if (ip_tags_datasource.filename().empty()) {
+    creation_status = absl::InvalidArgumentError("Cannot load tags from empty file path.");
+    return;
+  }
+  if (ip_tags_refresh_interval_ms_ > std::chrono::milliseconds(0) &&
+      ip_tags_datasource.has_watched_directory()) {
+    needs_refresh_ = true;
+  }
+  tags_ = tags_loader_.loadTags(ip_tags_datasource, dispatcher, tls, creation_status);
+}
+
+IpTagsProvider::~IpTagsProvider() { ENVOY_LOG(debug, "Shutting down ip tags provider"); };
+
+LcTrieSharedPtr IpTagsProvider::ipTags() ABSL_LOCKS_EXCLUDED(ip_tags_mutex_) {
```

Comment:
I believe this is being called on the data-plane path which might be costly (depending on how long will it take to load the new tags file). This may be more problematic if multiple threads are racing to this location and invoking `ipTags()` and `refreshTags()` concurrently.
The typical way it is done in Envoy is decoupling the config and data planes paths.
The config-path can be invoked using a timer or when a watch directory happens, and it will load the new file (on the main thread). Once it is done it atomically replaces the shared pointer that the data-path will fetch during the filter instantiation.
My motivation here is to decouple the config loading from the usage of the ip-tags, and ensure that the ip-tags are available even if they are updated during the filter's execution.

Reply:
After further inspecting the code, datasource reload itself happens on main thread and ip tags provider (which is a singleton) will get a tls copy of the ip tags string. The expensive part here is parsing of a new string into trie like structure (since ip tags file can be large in size). This part can be moved to a custom thread that runs a timer and periodically reparses the structure and swaps the pointer in provider from old data to new data (trie like structure).

Reply:
SGTM, will file separate PR for it